go/kbcmf/encrypt.go

```diff
@@ -0,0 +1,226 @@
+
+package kbcmf
+
+import (
+	"bytes"
+	"encoding/hex"
+	"golang.org/x/crypto/nacl/secretbox"
+	"io"
+)
+
+type publicEncryptStream struct {
+	output     io.Writer
+	header     *EncryptionHeader
+	sessionKey SymmetricKey
+	buffer     bytes.Buffer
+	inblock    []byte
+	macGroups  []SymmetricKey
+
+	numBlocks encryptionBlockNumber // the lower 64 bits of the nonce
+
+	didHeader bool
+	eof       bool
+	err       error
+}
+
+func (pes *publicEncryptStream) Write(plaintext []byte) (int, error) {
+
+	if !pes.didHeader {
+		pes.didHeader = true
+		pes.err = encodeNewPacket(pes.output, pes.header)
+	}
+
+	if pes.err != nil {
+		return 0, pes.err
+	}
+
+	var ret int
+	if ret, pes.err = pes.buffer.Write(plaintext); pes.err != nil {
+		return 0, pes.err
+	}
+	for pes.buffer.Len() >= EncryptionBlockSize {
+		pes.err = pes.encryptBlock()
+		if pes.err != nil {
+			return 0, pes.err
+		}
+	}
+	return ret, nil
+}
+
+func (pes *publicEncryptStream) macForAllGroups(b []byte) [][]byte {
+	var macs [][]byte
+	for _, key := range pes.macGroups {
+		mac := hmacSHA512(key[:], b)
+		macs = append(macs, mac)
+	}
+	return macs
+}
+
+func (pes *publicEncryptStream) encryptBlock() error {
+	var n int
+	var err error
+	n, err = pes.buffer.Read(pes.inblock[:])
+	if err != nil {
+		return nil
+	}
+	return pes.encryptBytes(pes.inblock[0:n])
+}
+
+func (pes *publicEncryptStream) encryptBytes(b []byte) error {
+
+	if err := pes.numBlocks.check(); err != nil {
+		return err
+	}
+
+	nonce := pes.numBlocks.newCounterNonce()
+	ciphertext := secretbox.Seal([]byte{}, b, (*[24]byte)(nonce), (*[32]byte)(&pes.sessionKey))
+	// Compute the MAC over the nonce and the ciphertext
+	sum, err := hashCryptoBlock(nonce, ciphertext)
+	if err != nil {
+		return err
+	}
+	macs := pes.macForAllGroups(sum)
+	block := EncryptionBlock{
+		Version:    PacketVersion1,
+		Tag:        PacketTagEncryptionBlock,
+		Ciphertext: ciphertext,
+		MACs:       macs,
+	}
+
+	if err = encodeNewPacket(pes.output, block); err != nil {
+		return nil
+	}
+
+	pes.numBlocks++
+	return nil
+}
+
+func (pes *publicEncryptStream) init(sender BoxSecretKey, receivers [][]BoxPublicKey) error {
+	eh := &EncryptionHeader{
+		Version:   PacketVersion1,
+		Tag:       PacketTagEncryptionHeader,
+		Sender:    sender.GetPublicKey().ToKID(),
+		Receivers: make([]receiverKeysCiphertext, 0, len(receivers)),
+	}
+	pes.header = eh
+	if err := randomFill(pes.sessionKey[:]); err != nil {
+		return err
+	}
+
+	// Only fill the first 20 bytes of the nonce. The remaining 4
+	// we'll increment with every call to Box
+	nonceRandLen := 20
+	var nonce Nonce
+	if err := randomFill(nonce[:nonceRandLen]); err != nil {
+		return err
+	}
+
+	// We don't necessarily have to copy our header nonce into place,
+	// but if feels safer, since we modify the nonce below
+	eh.Nonce = make([]byte, nonceRandLen)
+	copy(eh.Nonce, nonce[:])
+
+	d := make(map[string]struct{})
+
+	var i uint32
+
+	for gid, group := range receivers {
+		var macKey SymmetricKey
```

Comment:
I wonder if it makes sense to have the mac keys just be 128-bit. We're truncating the digests that short after all.